java/ql/src/experimental/Security/CWE/CWE-348/UseOfLessTrustedSourceLib.qll

```diff
@@ -0,0 +1,88 @@
+import java
+import DataFlow
+import semmle.code.java.dataflow.DataFlow
+import experimental.semmle.code.java.Logging
+
+/** A data flow source of the value of the `x-forwarded-for` field in the `header`. */
+class UseOfLessTrustedSource extends DataFlow::Node {
+  UseOfLessTrustedSource() {
+    exists(MethodAccess ma |
+      ma.getMethod().hasName("getHeader") and
+      ma.getArgument(0).toString().toLowerCase() = "\"x-forwarded-for\"" and
+      ma = this.asExpr()
+    )
+  }
+}
+
+/** A data flow sink of method return or log output or local print. */
+class UseOfLessTrustedSink extends DataFlow::Node {
+  UseOfLessTrustedSink() {
+    exists(ReturnStmt rs | rs.getResult() = this.asExpr())
```

Comment:
If we're treating returning as a sink we might as well flag any use of X-Forwarded-For except splitting and taking the rightmost value.

Reply:
The result of `return` as a sink, I think the method is a util method, this usage is common.

Reply:
Right, but you don't care about what is done with it subsequently? This is so general that there's almost no point checking how it is used at all.

Reply:
If so, I think there is a problem as long as there is a place to call this method. If you consider what to do with the result of the client ip method, I think there will be a false negative rate. What do you think?

Reply:
What uses do you actually consider dangerous, besides logging?

Reply:
In addition to local output or log records, there are also user authentication based on client ip, database storage, etc., but their use is not unique, so I think it is impossible to standardize the definition.

Reply:
Can we let the security-lab review it together?

Reply:
E.g:
- The client ip detects whether there is login permission: [UserServiceImpl](https://github.com/dinhlong1210/SpringSecurity/blob/550094117bd84690a9dee1e63c4605a60eae45e0/src/main/java/com/github/tutorial/Service/UserServiceImpl.java#L39)
- Limit the number of logins for the client ip, which is common in scenarios such as blasting and ticketing: [CaptchaService](https://github.com/dinhlong1210/SpringSecurity/blob/550094117bd84690a9dee1e63c4605a60eae45e0/src/main/java/com/github/tutorial/captcha/CaptchaService.java#L35)
- Insert the database as a record: [LogDB](https://github.com/aabouda/3ATechnologies/blob/a3c077be620adfed2dd3697b394e48e9c7172757/src/main/java/com/controllers/LogDB.java#L31)
- Save it in `JavaBean` and insert into the database: [MenuController](https://github.com/5202m/chatadmin/blob/dbaafbfba673172d1d5286b703e68536866d3897/src/main/java/com/gwghk/mis/controller/MenuController.java#L106)

Reply:
I can, but as this stands I would have to warn them that the query will produce really strange results:

* Directly used in logging call: positive
* Directly used in a rate-limiter: negative
* Directly returned: positive
* Returned in a field: negative

A user of this query would be very confused, because their expectation is that the way the value is used is the important thing, not trivial aspects of dataflow like is it returned from a function or used in function-local scope.

I would recommend you find some e.g. rate-limiting utility that would constitute a dangerous way to use this identifier. Alternatively you can choose to send this to the security lab with the health warning attached.

Reply:
Thank you, I agree to let them know the current situation, and I also want to listen to their suggestions, so as to implement a changed query.